cmd2/AutoCompleter.py

```diff
@@ -0,0 +1,796 @@
+                            flag_action = self._flag_to_action[token]
+                        elif hasattr(self._parser, 'allow_abbrev') and self._parser.allow_abbrev:
+                            candidates_flags = [flag for flag in self._flag_to_action.keys() if flag.startswith(token)]
+                            if len(candidates_flags) == 1:
+                                flag_action = self._flag_to_action[candidates_flags[0]]
+
+                        if flag_action is not None:
+                            # resolve argument counts
+                            self._process_action_nargs(flag_action, flag_arg)
+                            if not is_last_token and not isinstance(flag_action, argparse._AppendAction):
+                                matched_flags.extend(flag_action.option_strings)
+
+                    # current token isn't a potential flag
+                    #   - does the last flag accept variable arguments?
+                    #   - have we reached the max arg count for the flag?
+                    elif not flag_arg.variable or flag_arg.count >= flag_arg.max:
+                        # previous flag doesn't accept variable arguments, count this as a positional argument
+
+                        # reset flag tracking variables
+                        flag_arg.reset()
+                        flag_action = None
+                        current_is_positional = True
+
+                        if len(token) > 0 and pos_action is not None and pos_arg.count < pos_arg.max:
+                            # we have positional action match and we haven't reached the max arg count, consume
+                            # the positional argument and move on.
+                            consume_positional_argument()
+                        elif pos_action is None or pos_arg.count >= pos_arg.max:
+                            # if we don't have a current positional action or we've reached the max count for the action
+                            # close out the current positional argument state and set up for the next one
+                            pos_index = next_pos_arg_index
+                            next_pos_arg_index += 1
+                            pos_arg.reset()
+                            pos_action = None
+
+                            # are we at a sub-command? If so, forward to the matching completer
+                            if pos_index < len(self._positional_actions):
+                                action = self._positional_actions[pos_index]
+                                pos_name = action.dest
+                                if pos_name in self._positional_completers.keys():
+                                    sub_completers = self._positional_completers[pos_name]
+                                    if token in sub_completers.keys():
+                                        return sub_completers[token].complete_command(tokens, text, line,
```

Comment:
This code near here-ish has a whole lot of nested conditionals.  At his point I think we are 7 or 8 levels deep.  I'm not entirely sure how to make it less nested.  Just something to think about.  It just makes it somewhat difficult to follow the logic.

Reply:
Yeah, I think it would actually be harder to follow if I break this up differently. I tried to comment clearly what I'm trying to do here. Hopefully that helps.